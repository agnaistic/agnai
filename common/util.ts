import { AppSchema } from './types/schema'

export function replace<T extends { _id: string }>(id: string, list: T[], item: Partial<T>) {
  return list.map((li) => (li._id === id ? { ...li, ...item } : li))
}

export function findOne<T extends { _id: string }>(id: string, list: T[]): T | void {
  for (const item of list) {
    if (item._id === id) return item
  }
}

export function toArray<T>(values?: T | T[]): T[] {
  if (values === undefined) return []
  if (Array.isArray(values)) return values
  return [values]
}

export function distinct<T extends { _id: string }>(values: T[]) {
  const set = new Set<string>()
  const next: T[] = []

  for (const item of values) {
    if (set.has(item._id)) continue
    next.push(item)
    set.add(item._id)
  }

  return next
}
export function wait(secs: number) {
  return new Promise((res) => setTimeout(res, secs * 1000))
}

export function waitMs(ms: number) {
  return new Promise((res) => setTimeout(res, ms))
}

export function isObject(val: unknown) {
  return Object.prototype.toString.call(val) === '[object Object]'
}

export function cycleArray<T>(arr: T[], shiftAmount: number) {
  const effectiveShift = shiftAmount % arr.length
  return [...arr.slice(effectiveShift), ...arr.slice(0, effectiveShift)]
}

export const dateStringFromUnix = (unixTimestamp: number): string =>
  new Date(unixTimestamp).toISOString()

export function uniqBy<T, U>(arr: T[], accessor: (el: T) => U) {
  return arr.flatMap((el, i) => (arr.slice(0, i).map(accessor).includes(accessor(el)) ? [] : [el]))
}

export function toDuration(valueSecs: number, full?: boolean) {
  if (valueSecs === 0) {
    if (full) return '0s'
    return '0 seconds'
  }
  const {
    duration: [days, hours, minutes, seconds],
  } = toRawDuration(valueSecs)

  if (full) {
    return [`${days}d`, `${hours}h`, `${minutes}m`, `${seconds}s`]
      .filter((time) => !time.startsWith('0'))
      .join(':')
  }

  if (days) {
    return `${days} ${days > 1 ? 'days' : 'day'}`
  }

  if (hours) {
    return `${hours} ${hours > 1 ? 'hours' : 'hour'}`
  }

  if (minutes) {
    return `${minutes} ${minutes > 1 ? 'mins' : 'min'}`
  }

  return `${seconds} ${seconds > 1 ? 'seconds' : 'second'}`
}

export function elapsedSince(date: string | Date, offsetMs: number = 0) {
  if (!date) return 'unknown'
  const diff = Date.now() - new Date(date).valueOf() + offsetMs
  const elapsed = Math.floor(diff / 1000)
  if (elapsed < 60) return 'a moment'
  return toDuration(Math.floor(elapsed))
}

const ONE_HOUR = 3600
const ONE_DAY = 86400

type Days = number
type Hours = number
type Minutes = number
type Seconds = number

type Duration = [Days, Hours, Minutes, Seconds]

function toRawDuration(valueSecs: number) {
  const days = Math.floor(valueSecs / ONE_DAY)
  const hours = Math.floor(valueSecs / ONE_HOUR) % 24
  const mins = Math.floor(valueSecs / 60) % 60
  const secs = Math.ceil(valueSecs % 60)

  return {
    duration: [days, hours, mins, secs] as Duration,
    seconds: valueSecs,
    text: valueSecs <= 0 ? 'now' : `${days}d:${hours}h:${mins}m:${secs}s`.replace('0d:', ''),
  }
}

/**
 * Remove leading tabs and whitespace on each line from template strings
 * Also 'trim' the text before returning
 */
export function neat(params: TemplateStringsArray, ...rest: string[]) {
  let str = ''
  for (let i = 0; i < params.length; i++) {
    str += params[i] + (rest[i] || '')
  }

  return str
    .split('\n')
    .map((line) => line.replace(/^[\t ]+/g, ''))
    .join('\n')
    .trim()
}

const end = `."'*!?)}]\``.split('')
export function trimSentence(text: string) {
  const last = end.reduce((last, char) => {
    const index = text.lastIndexOf(char)
    return index > last ? index : last
  }, -1)

  if (last === -1) return text
  return text.slice(0, last + 1)
}

export function slugify(str: string) {
  return str
    .toLowerCase()
    .trim()
    .replace(/[^\w\s-]/g, '')
    .replace(/[\s_-]+/g, '-')
    .replace(/^-+|-+$/g, '')
}

// https://stackoverflow.com/a/3561711
export function escapeRegex(string: string) {
  return string.replace(/[/\-\\^$*+?.()|[\]{}]/g, '\\$&')
}

export function getBotName(
  chat: AppSchema.Chat,
  msg: AppSchema.ChatMessage,
  chars: Record<string, AppSchema.Character>,
  replyAs: AppSchema.Character,
  main: AppSchema.Character,
  impersonate?: AppSchema.Character
) {
  if (!msg.characterId) return replyAs?.name || main.name
  if (msg.characterId.startsWith('temp-')) {
    const temp = chat.tempCharacters?.[msg.characterId]
    if (!temp) return main.name
    return temp.name
  }

  const char =
    msg.characterId && impersonate?._id === msg.characterId ? impersonate : chars[msg.characterId]

  if (!char) {
    return main.name
  }

  return char.name
}

export function eventGenerator<T = any>() {
  const queue: Array<Promise<any>> = []
  let done = false
  let signal = false

  const stream = (async function* () {
    while (!done) {
      await waitMs(1)
      const promise = queue.shift()
      if (!promise) {
        if (signal) {
          done = true
        }
        continue
      }

      const result = await promise
      yield result
    }
  })() as AsyncGenerator<T, T>

  return {
    stream,
    push: (value: T) => {
      if (done) return
      queue.push(Promise.resolve(value))
    },
    done: () => {
      if (done) return
      signal = true
    },
  }
}

/**
 * Will automatically clamp to `-max -> max` if `min` is not provided
 */
export function clamp(toClamp: number, max: number, min?: number) {
  return Math.max(Math.min(toClamp, max), min ?? -max)
}

export function now() {
  return new Date().toISOString()
}
