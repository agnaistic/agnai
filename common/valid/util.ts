import { OptionalPrimitive, Primitive, Validator } from './types'

export function isPrimitive(value: any): value is Primitive {
  return (
    typeof value === 'string' &&
    (value === 'string' ||
      value === 'boolean' ||
      value === 'number' ||
      value === 'any' ||
      value === 'unknown')
  )
}

export function isOptionalPrimitive(value: any): value is OptionalPrimitive {
  return (
    value === 'string?' ||
    value === 'boolean?' ||
    value === 'number?' ||
    value === 'any?' ||
    value === 'unknown?'
  )
}

export function isTuplePrimitive(value: any): value is [Primitive] | readonly [Primitive] {
  if (Array.isArray(value) === false) return false
  if (value.length !== 1) return false
  if (!isPrimitive(value[0])) return false
  return true
}

export function isTupleOptional(
  value: any
): value is [OptionalPrimitive] | readonly [OptionalPrimitive] {
  if (Array.isArray(value) === false) return false
  if (value.length !== 1) return false
  if (!isOptionalPrimitive(value[0])) return false
  return true
}

export function isOptionalArray(value: any): value is [Validator, '?'] | readonly [Validator, '?'] {
  if (Array.isArray(value) === false) return false
  if (value.length !== 2) return false
  return value[1] === '?'
}

export function isObjectOptional(value: any): value is Validator & { '?': any } {
  return typeof value === 'object' && '?' in value && !!value['?']
}

export function isTupleBody(
  value: any
): value is [Validator] | readonly [Validator] | [Validator, '?'] | readonly [Validator, '?'] {
  if (Array.isArray(value) === false) return false
  if (value.length !== 1 && value.length !== 2) return false
  if (value.length === 2) return typeof value[0] === 'object' && value[1] === '?'
  return typeof value[0] === 'object'
}

export function isOptionalUnion<T extends string>(
  value: any
): value is T[] | readonly string[] | undefined {
  if (!Array.isArray(value)) return false
  if (value.length < 1) return false
  if (isPrimitive(value[0])) return false
  if (value.includes(null)) return true
  return false
}

export function isUnion<T extends string>(value: any): value is [T] | string[] | readonly string[] {
  if (Array.isArray(value) === false) return false
  if (value.length < 1) return false
  if (isPrimitive(value[0])) return false

  return true
}
