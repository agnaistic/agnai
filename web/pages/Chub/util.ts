import { CHUB_URL } from '/web/store/chub'

const headers = {
  'Content-Type': 'application/json',
  accept: '/',
}

export async function processBook(fullPath: string) {
  const body = { format: 'AGNAI', fullPath, version: 'main' }
  const res = await fetch(`${CHUB_URL}/lorebooks/download`, {
    headers,
    body: JSON.stringify(body),
    method: 'post',
  })

  const json = await res.json()
  return json
}

export async function processChar(fullPath: string) {
  const body = { format: 'tavern', fullPath, version: 'main' }
  const res = await fetch(`${CHUB_URL}/characters/download`, {
    headers,
    body: JSON.stringify(body),
    method: 'post',
  })

  const blob = await res.blob()
  const file = new File([blob], `main_${fullPath}.png`, { type: 'image/png' })
  return file
}
