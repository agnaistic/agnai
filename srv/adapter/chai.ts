import needle from 'needle'
import { config } from '../config'
import { trimResponse, trimResponseV2 } from '../api/chat/common'
import { ModelAdapter } from './type'

export const handleChai: ModelAdapter = async function* ({ char, members, prompt, settings, log }) {
  if (!config.chai.url) {
    yield { error: 'Chai URL not set' }
    return
  }

  if (config.chai.uid === 'empty' || config.chai.key === 'empty') {
    yield { error: 'Chai is not configured' }
    return
  }

  const body = { ...settings, text: prompt }

  log.debug(body, 'Chai payload')

  const response = await needle('post', `${config.chai.url}/generate/gptj`, body, {
    json: true,
    timeout: 3000,
    response_timeout: 10000,
    headers: {
      'user-agent':
        'Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/110.0.0.0 Safari/537.36',
      origin: 'https://chai.ml',
      developer_uid: config.chai.uid,
      developer_key: config.chai.key,
    },
  }).catch((err) => ({ err }))

  if ('err' in response) {
    yield { error: response.err.message }
    return
  }

  const status = response.statusCode || 0
  if (status >= 500) {
    log.error(response.body, 'Chai request failed')
    yield { error: `Chai internal server error: ` }
    return
  }

  if (status >= 400) {
    log.error(response.body, 'Chai request failed')
    yield { error: 'Chai request error' }
    return
  }

  const username = 'You'
  const endTokens = [
    `Me:`,
    `Me :`,
    `${username}:`,
    `${char.name}:`,
    `${username} :`,
    `${char.name} :`,
    'END_OF_DIALOG',
  ]

  const trimmed = trimResponseV2(response.body.data, char, members, endTokens)
  yield trimmed || response.body.data
}
