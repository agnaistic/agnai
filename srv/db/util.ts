import crypto from 'crypto'
import bcrypt from 'bcryptjs'
import { config } from '../config'
import { AppSchema } from '../../common/types/schema'
import { defaultChars } from '../../common/characters'

const ALGO = 'aes-192-cbc'
const KEY = crypto.scryptSync(config.jwtSecret, 'salt', 24)

export function now() {
  return new Date().toISOString()
}

export async function encryptPassword(value: string) {
  const salt = await bcrypt.genSalt()
  const hash = await bcrypt.hash(value, salt)

  return hash
}

export function encryptText(text: string) {
  const iv = crypto.randomBytes(16)
  const cipher = crypto.createCipheriv(ALGO, KEY, iv)
  const encrypted = cipher.update(text, 'utf8', 'hex')

  return [encrypted + cipher.final('hex'), Buffer.from(iv).toString('hex')].join('|')
}

export function decryptText(text: string) {
  const [encrypted, iv] = text.split('|')
  if (!iv) throw new Error('IV not found')
  const decipher = crypto.createDecipheriv(ALGO, KEY, Buffer.from(iv, 'hex'))
  return decipher.update(encrypted, 'hex', 'utf8') + decipher.final('utf8')
}

export const STARTER_CHARACTER: AppSchema.Character = {
  _id: '',
  userId: '',
  kind: 'character',
  createdAt: '',
  updatedAt: '',
  favorite: false,
  visualType: 'avatar',
  ...defaultChars.Robot,
}

export async function wait(ms: number) {
  return new Promise((res) => setTimeout(res, ms))
}
