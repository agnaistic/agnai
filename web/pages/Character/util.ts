import { AppSchema } from '/common/types'
import { getAssetUrl, storage } from '/web/shared/util'
import { toastStore } from '/web/store'
import { ALLOWED_TYPES, charsApi, getImageData } from '/web/store/data/chars'
import { exportCharacter } from '/common/characters'
import text from 'png-chunk-text'
import extract from 'png-chunks-extract'
import encode from 'png-chunks-encode'

const CACHE_KEY = 'agnai-chatlist-cache'

export type ChatCharacter = { _id: string; name: string; description: string; avatar?: string }

export type ChatLine = {
  _id: string
  characters: ChatCharacter[]
  name: string
  createdAt: string
  updatedAt: string
  characterId: string
}

export type SortType =
  | 'chat-updated'
  | 'chat-created'
  | 'character-name'
  | 'character-created'
  | 'bot-activity'

export type SortDirection = 'asc' | 'desc'

export type ChatGroup = { char: AppSchema.Character | null; chats: ChatLine[] }

export type ListCache = {
  sort: {
    field: SortType
    direction: SortDirection
  }
}

export function getChatSortableValue(chat: ChatLine, field: SortType) {
  switch (field) {
    case 'chat-updated':
      return chat.updatedAt
    case 'chat-created':
      return chat.createdAt
    default:
      return 0
  }
}

export function getChatSortFunction(field: SortType, direction: SortDirection) {
  return (left: ChatLine, right: ChatLine) => {
    const mod = direction === 'asc' ? 1 : -1
    const l = getChatSortableValue(left, field)
    const r = getChatSortableValue(right, field)
    return l > r ? mod : l === r ? 0 : -mod
  }
}

export function getCharacterSortableValue(char: AppSchema.Character, field: SortType) {
  switch (field) {
    case 'character-name':
      return char.name.toLowerCase()

    case 'character-created':
      return char.createdAt

    case 'bot-activity':
      return char.updatedAt

    default:
      return 0
  }
}

export function getCharSortFunction(type: SortType, direction: SortDirection) {
  return (left: AppSchema.Character, right: AppSchema.Character) => {
    const mod = direction === 'asc' ? 1 : -1
    const l = getCharacterSortableValue(left, type)
    const r = getCharacterSortableValue(right, type)
    return l > r ? mod : l === r ? 0 : -mod
  }
}

export function groupAndSort(
  allChars: AppSchema.Character[],
  allChats: ChatLine[],
  type: SortType,
  direction: SortDirection
): Array<ChatGroup> {
  if (type === 'chat-updated' || type === 'chat-created') {
    const sorted = allChats.slice().sort(getChatSortFunction(type, direction))
    return [{ char: null, chats: sorted }]
  }

  const groups: ChatGroup[] = []
  const sortedChats = allChats.slice().sort(getChatSortFunction('chat-updated', 'desc'))

  const chars = allChars.slice().map((char) => {
    if (type !== 'bot-activity') return char
    const first = sortedChats.find((c) => c.characters.some((ch) => ch._id === char._id))
    return { ...char, updatedAt: first?.updatedAt || new Date(0).toISOString() }
  })

  chars.sort(getCharSortFunction(type, direction))

  for (const char of chars) {
    const chats = allChats
      .filter((ch) => ch.characters.some((c) => c._id === char._id))
      .sort(getChatSortFunction('chat-updated', 'desc'))
    groups.push({ char, chats })
  }

  return groups
}

export function getListCache(): ListCache {
  const existing = storage.localGetItem(CACHE_KEY)
  const defaultCache: ListCache = { sort: { field: 'chat-updated', direction: 'desc' } }

  if (!existing) {
    return defaultCache
  }

  return { ...defaultCache, ...JSON.parse(existing) }
}

export function saveListCache(cache: ListCache) {
  storage.localSetItem(CACHE_KEY, JSON.stringify(cache))
}

export function toCharacterMap(bots: AppSchema.Character[]) {
  const map = bots.reduce<Record<string, AppSchema.Character>>(
    (prev, curr) => Object.assign(prev, { [curr._id]: curr }),
    {}
  )
  return map
}

export async function convertCharToJSON(
  char: AppSchema.Character | string,
  format: string,
  schema: string
) {
  if (typeof char === 'string') {
    const res = await charsApi.getCharacterDetail(char)
    if (res.error) {
      return toastStore.error(`Failed to download character: ${res.error}`)
    }

    if (res.result) {
      const json = charToJson(res.result, format, schema)
      return `data:text/json:charset=utf-8,${encodeURIComponent(json)}`
    }
  } else {
    const json = charToJson(char, format, schema)
    return `data:text/json:charset=utf-8,${encodeURIComponent(json)}`
  }
}

export async function downloadCharCard(
  input: string | AppSchema.Character,
  format: string,
  schema: string
) {
  let char: AppSchema.Character

  if (typeof input === 'string') {
    const res = await charsApi.getCharacterDetail(input)
    if (res.error) {
      return toastStore.error(`Failed to download character: ${res.error}`)
    } else {
      char = res.result!
    }
  } else {
    char = input
  }

  const json = charToJson(char, format, schema)
  const image = getAssetUrl(char.avatar!)
  /**
   * Only PNG and APNG files can contain embedded character information
   * If the avatar image is not either of these formats, we must convert it
   */

  const dataurl = await imageToDataURL(image)
  const base64 = dataurl.split(',')[1]
  const imgBuffer = Buffer.from(window.atob(base64), 'binary')
  const chunks = extract(imgBuffer).filter((chunk) => chunk.name !== 'tEXt')
  const output = Buffer.from(json, 'utf8').toString('base64')
  const lastChunkIndex = chunks.length - 1
  const chunksToExport = [
    ...chunks.slice(0, lastChunkIndex),
    text.encode('chara', output),
    chunks[lastChunkIndex],
  ]
  const anchor = document.createElement('a')
  anchor.href = URL.createObjectURL(new Blob([Buffer.from(encode(chunksToExport))]))
  anchor.download = `${char.name}.card.png`
  anchor.click()
  URL.revokeObjectURL(anchor.href)
}

async function imageToDataURL(image: string) {
  const { ext } = getExt(image)

  const base64 = await getImageBase64(image)
  const apng = await isAPNG(base64)
  if (apng || ext === 'apng') {
    return base64
  }

  const element = await asyncImage(base64)

  const canvas = document.createElement('canvas')
  canvas.width = element.image.naturalWidth
  canvas.height = element.image.naturalHeight
  const ctx = canvas.getContext('2d')
  ctx?.drawImage(element.image, 0, 0)
  const dataUrl = canvas.toDataURL('image/png')
  return dataUrl
}

function charToJson(char: AppSchema.Character, format: string, schema: string) {
  const { _id, ...json } = char

  const copy = { ...char }
  copy.persona.kind = schema as any

  if (format === 'native') {
    return JSON.stringify(json, null, 2)
  }

  const content = exportCharacter(copy, format as any)
  return JSON.stringify(content, null, 2)
}

function getExt(url: string): { type: 'base64' | 'url'; ext: string } {
  if (url.startsWith('data:')) {
    const [header] = url.split(',')
    const ext = header.slice(11, -7)
    return ALLOWED_TYPES.has(ext) ? { type: 'base64', ext } : { type: 'base64', ext: 'unknown' }
  }

  const ext = url.split('.').slice(-1)[0]
  if (ALLOWED_TYPES.has(ext)) return { type: 'url', ext }
  return { type: 'url', ext: 'unknown' }
}

async function getImageBase64(image: string) {
  if (image.startsWith('data:')) return image

  if (!image.startsWith('http')) {
    image = getAssetUrl(image)
  }

  const base64 = await getImageData(image)
  return base64!
}

function asyncImage(src: string) {
  return new Promise<{ name: string; image: HTMLImageElement }>(async (resolve, reject) => {
    const data = await getImageBase64(src)
    const image = new Image()
    image.setAttribute('crossorigin', 'anonymous')
    image.src = data

    image.onload = () => resolve({ name: src, image })
    image.onerror = (ev) => reject(ev)
  })
}

async function isAPNG(base64: string) {
  if (base64.startsWith('data:')) {
    base64 = base64.split(',')[1]
  }
  const buffer = Buffer.from(window.atob(base64), 'binary')
  try {
    for (const chunk of extract(buffer)) {
      if (chunk.name === 'IDAT') return false
      if (chunk.name === 'acTL') return true
    }

    return false
  } catch (ex) {
    return false
  }
}
