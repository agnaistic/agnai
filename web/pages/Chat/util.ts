import { AppSchema } from '/common/types'
import { settingStore, userStore } from '/web/store'

/**
 * Retrieve the unique set of active bots for a conversation
 */
export function getActiveBots(
  chat: AppSchema.Chat,
  bots: Record<string, AppSchema.Character>,
  exclude?: Record<string, any>
) {
  if (!chat) return []
  const unique = new Set([chat.characterId])

  for (const [id, active] of Object.entries(chat.characters || {})) {
    if (exclude && id in exclude) continue
    if (!active) continue
    if (!bots[id]) continue
    unique.add(id)
  }

  for (const [id, active] of Object.entries(chat.tempCharacters || {})) {
    if (exclude && id in exclude) continue
    if (!active) continue
    if (active.favorite === false) continue
    if (active.deletedAt) continue
    unique.add(id)
  }

  const all = Array.from(unique)
    .map((id) => bots[id] || chat.tempCharacters?.[id])
    .filter((bot) => !!bot)
    .sort(tempSort)
  return all
}

export function getBotsForChat(
  chat: AppSchema.Chat,
  main: AppSchema.Character,
  bots: Record<string, AppSchema.Character>
) {
  const chars: Record<string, AppSchema.Character> = { [main._id]: main }

  for (const key in chat.characters || {}) {
    const bot = bots[key]
    if (bot) chars[key] = bot
  }

  for (const [id, char] of Object.entries(chat.tempCharacters || {})) {
    chars[id] = char
  }

  return chars
}

export function canConvertGaslightV2(preset: Partial<AppSchema.UserGenPreset>) {
  // Do not upgrade built-in presets
  if (!preset?._id) return false

  // Do not upgrade presets without a gaslight
  if (!preset.gaslight) return false

  // Do not upgrade presets with an unknown service -- only models that support instruct are supported
  if (!preset.service) return false

  return true
}

function tempSort(a: AppSchema.Character, b: AppSchema.Character) {
  return +!b._id.startsWith('temp-') - +!a._id.startsWith('temp-') || a.name.localeCompare(b.name)
}

export function isEligible() {
  const cfg = settingStore.getState()
  const user = userStore.getState()

  const userLevel = user.user?.sub?.level ?? -1
  const eligible = cfg.config.subs.some((sub) => userLevel >= sub.level)

  return eligible
}
