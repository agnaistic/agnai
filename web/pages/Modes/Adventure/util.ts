import { v4 } from 'uuid'
import { gameStore } from './state'
import { downloadJson } from '/web/shared/util'
import { toastStore } from '/web/store'
import { GuidedTemplate } from '/web/store/data/guided'

const emptyTemplate: GuidedTemplate = {
  _id: '',
  name: '',
  byline: '',
  description: '',

  display: '',
  response: '',

  init: '',
  loop: '',
  history: '',
  introduction: '',

  fields: [],
  lists: {},
  manual: [],
}

export function exportTemplate(id: string) {
  const { templates } = gameStore.getState()
  const template = templates.find((t) => t._id === id)

  if (!template) {
    toastStore.error(`Export failed: Could not find template`)
    return
  }

  downloadJson(template, `template_${id.slice(0, 4)}`)
}

export function validateTemplate(template: any): asserts template is GuidedTemplate {
  const missing: string[] = []
  for (const key of Object.keys(emptyTemplate)) {
    if (key in template === false) {
      missing.push(key)
    }
  }

  if (missing.length) {
    throw new Error(`Invalid template, missing keys: ${missing.join(', ')}`)
  }
}

export function importTemplate(template: any) {
  validateTemplate(template)

  const next = {
    ...template,
    _id: v4(),
  }

  gameStore.importTemplate(next)
  return next
}
