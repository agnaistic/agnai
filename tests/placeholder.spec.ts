import { expect } from 'chai'
import './init'
import { reset, template } from './util'

describe('Placeholder tests', () => {
  before(reset)

  it('will not include duplicates in {{all_personalities}}', () => {
    const actual = template(`{{personality}}\n\n{{all_personalities}}`, {})

    expect(actual).toMatchSnapshot()
  })
})
