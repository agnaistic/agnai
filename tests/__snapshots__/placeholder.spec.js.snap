// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Placeholder tests will not include duplicates in {{all_personalities}} 1`] = `
Object {
  "inserts": Map {},
  "parsed": "OtherBot replies a lot

MainChar's personality: MainCharacter talks a lot",
}
`;
